### PR TITLE
`StaticRoute`: Enable `auth` option

### PR DESCRIPTION
**Pull request by mbana**
Resolves kubeshop/kusk-gateway/issues/763.

See: <https://github.com/kubeshop/kusk-gateway/issues/763>.

Closing <https://github.com/kubeshop/kusk-gateway/pull/855> in favour of this one.

Signed-off-by: Mohamed Bana <mohamed@bana.io>

### CODE REVIEW

File: internal/controllers/config_manager.go [hunk 1]
```diff
@@ -143,14 +144,15 @@ func (c *KubeEnvoyConfigManager) UpdateConfiguration(ctx context.Context, fleetI
 			return fmt.Errorf("failed to generate options from the static route config: %w", err)
 		}
 
-		if err := UpdateConfigFromOpts(envoyConfig, opts); err != nil {
-			return fmt.Errorf("failed to generate config: %w", err)
+		kubernetesClient := c.Client
```
Comment (jasmingacic) [inline internal/controllers/config_manager.go:147]: why make a copy here in the first place when it can be passed to the function as is

File: internal/controllers/parser.go [hunk 2]
```diff
@@ -82,7 +83,7 @@ func UpdateConfigFromAPIOpts(
 	opts *options.Options,
 	spec *openapi3.T,
 	httpConnectionManagerBuilder *config.HCMBuilder,
-	clBuilder *cloudentity.Builder,
+	cloudEntityBuilder *cloudentity.Builder,
```
Comment (jasmingacic) [inline internal/controllers/parser.go:86]: I like that you made the code a little more readable

File: internal/controllers/parser.go [hunk 3]
```diff
@@ -495,9 +499,48 @@ func extractParams(parameters openapi3.Parameters) map[string]types.ParamSchema
 }
 
 // UpdateConfigFromOpts updates Envoy configuration from Options only
-func UpdateConfigFromOpts(envoyConfiguration *config.EnvoyConfiguration, opts *options.StaticOptions) error {
+func UpdateConfigFromOpts(
```
Comment (jasmingacic) [inline internal/controllers/parser.go:502]: thanks for doing this

File: internal/envoy/auth/oauth2_filter.go [hunk 4]
```diff
@@ -150,15 +153,15 @@ func NewFilterHTTPOAuth2(oauth2Options *options.OAuth2, args *parseAuthOptionsAr
...
-	// Disable OAuth2 on the root path - "/" - see: <https://github.com/kubeshop/kusk-gateway/issues/680>.
-	passThroughMatcher := []*envoy_config_route_v3.HeaderMatcher{
-		{
-			Name: ":path",
-			HeaderMatchSpecifier: &envoy_config_route_v3.HeaderMatcher_ExactMatch{
-				ExactMatch: "/",
-			},
-		},
-	}
+	// // Disable OAuth2 on the root path - "/" - see: <https://github.com/kubeshop/kusk-gateway/issues/680>.
```
Comment (jasmingacic) [inline internal/envoy/auth/oauth2_filter.go:156]: make sure to remove this once you are done

**Review by jasmingacic — CHANGES_REQUESTED** [review]
minor remarks but it is looking good

File: internal/envoy/auth/parser.go [hunk 5]
```diff
@@ -46,15 +46,14 @@ type CloudEntityBuilderArguments struct {
...
 	EnvoyConfiguration           *config.EnvoyConfiguration
 	HTTPConnectionManagerBuilder *config.HCMBuilder
 	CloudEntityBuilderArguments  *CloudEntityBuilderArguments
 	CloudEntityBuilder           *cloudentity.Builder
 	GenerateClusterName          generateClusterNameFunc
 	KubernetesClient             client.Client
-	PathsToIgnore                []string
 }
 
 func NewParseAuthOptionsArguments(
```
Comment (jasmingacic) [inline internal/envoy/auth/parser.go:59]: this has become pointless as we can achieve the same by initializing the variable instead of using constructor

**Review by jasmingacic — APPROVED** [review]
other than a minor comment it is looking good

File: internal/controllers/config_manager.go (hunk 1 above)
Comment (mbana) [reply internal/controllers/config_manager.go:146]: Resolved.

File: internal/envoy/auth/oauth2_filter.go (hunk 4 above)
Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go]: I've left it in as a comment. It's just useful to know why OAuth2 is being disabled on particular routes.

Hope that is OK?

File: internal/envoy/auth/parser.go (hunk 5 above)
Comment (mbana) [reply internal/envoy/auth/parser.go]: Resolved.

Removed `NewParseAuthOptionsArguments` constructor.

File: internal/controllers/config_manager.go [hunk 6]
```diff
@@ -111,8 +111,7 @@ func (c *KubeEnvoyConfigManager) UpdateConfiguration(ctx context.Context, fleetI
 			return fmt.Errorf("failed to validate options: %w", err)
 		}
 
-		kubernetesClient := c.Client
-		if err = UpdateConfigFromAPIOpts(envoyConfig, c.Validator, opts, apiSpec, httpConnectionManagerBuilder, cloudEntityBuilder, api.Name, kubernetesClient); err != nil {
+		if err = UpdateConfigFromAPIOpts(envoyConfig, c.Validator, opts, apiSpec, httpConnectionManagerBuilder, cloudEntityBuilder, api.Name, c.Client); err != nil {
```
Comment (jasmingacic) [inline internal/controllers/config_manager.go:114]: good that you cleaned it up

File: build/manager/Dockerfile [hunk 7]
```diff
@@ -31,7 +31,7 @@ RUN mkdir -m 00755 /opt/manager
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
 FROM --platform=$BUILDPLATFORM gcr.io/distroless/static:nonroot
 
-ENV GOTRACEBACK=all
+ENV GOTRACEBACK=single
```
Comment (kylehodgetts) [inline build/manager/Dockerfile:34]: What does this do?

File: examples/auth/oauth2/static-route/Makefile [hunk 8]
```diff
@@ -0,0 +1,6 @@
+.PHONY: all
+all:
+	-kubectl delete -f ./manifests.yaml
+	-kubectl delete -f ./api.yaml
+	-kubectl apply -f ./manifests.yaml
+	-kubectl apply -f ./api.yaml
```
Comment (kylehodgetts) [inline examples/auth/oauth2/static-route/Makefile:6]: Do we need a makefile for this? is this common that we will need to do this?

File: skaffold.sh [hunk 9]
```diff
@@ -1,27 +1,47 @@
...
-# To disable metrics: development/minikube.sh
-# To enable metrics: METRICS=1 development/minikube.sh
 set -o errexit  # Used to exit upon error, avoiding cascading errors
 set -o pipefail # Unveils hidden failures
 set -o nounset  # Exposes unset variables
 
 PROFILE="${PROFILE:-kgw}"
 
-mkdir -pv /tmp/skaffold || echo '`/tmp/skaffold` already exist - skipping create'
+check_cluster_is_running() {
```
Comment (kylehodgetts) [inline skaffold.sh:8]: This stuff should probably be in a different PR. It's causing noise here

**Review by kylehodgetts — CHANGES_REQUESTED** [review]

File: build/manager/Dockerfile (hunk 7 above)
Comment (mbana) [reply build/manager/Dockerfile:34]: For the change to `GOTRACEBACK=single` see <https://skaffold.dev/docs/workflows/debug/#go-runtime-go>.

File: examples/auth/oauth2/static-route/Makefile (hunk 8 above)
Comment (mbana) [reply examples/auth/oauth2/static-route/Makefile:6]: I'll remove  the `Makefile`.

File: internal/envoy/auth/oauth2_filter.go [hunk 10]
```diff
@@ -45,17 +45,19 @@ import (
 	"github.com/kubeshop/kusk-gateway/pkg/options"
 )
 
-func NewFilterHTTPOAuth2(oauth2Options *options.OAuth2, args *parseAuthOptionsArguments) (*anypb.Any, error) {
+func NewFilterHTTPOAuth2(oauth2Options *options.OAuth2, args *ParseAuthArguments) (*anypb.Any, *ParseAuthOutput, error) {
```
[the remark this answers is not in the document]
Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go:48]: What change do you have in mind?

[the remark this answers is not in the document]
Comment (kylehodgetts) [reply internal/envoy/auth/oauth2_filter.go:48]: The function name states it returns a new Oauth http filter. It should do just that.

If you need output, you should use a different function for that

File: skaffold.sh (hunk 9 above)
Comment (mbana) [reply skaffold.sh:8]: Yes, I agree.

It's not effecting anything though at the moment, so if it's OK, I'll just leave `skaffold.sh` as-is.

File: internal/envoy/auth/oauth2_filter.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go]: Please see latest commit and mark as resolved, if you're OK with it.

**Review by kylehodgetts — APPROVED** [review]

**Review by kylehodgetts — APPROVED** [review]
